### PR TITLE
Revise reserved & built in names list

### PR DESCRIPTION
**Pull request by garyb**
@michaelficarra could you take a look at this please?

### CODE REVIEW

**Comment by garyb** [comment]
Bump! /cc @michaelficarra

File: src/Language/PureScript/CodeGen/JS/Common.hs [hunk 1]
```diff
@@ -75,107 +81,160 @@ identCharToString c = '$' : show (ord c)
...
+  concat
+    [ jsReserved
+    , jsFutureReserved
+    , jsFutureReservedStrict
+    , jsOldReserved
+    , jsLiterals
+    ]
+
+jsReserved :: [String]
+jsReserved =
```
Comment (michaelficarra) [inline src/Language/PureScript/CodeGen/JS/Common.hs]: Can you rename this list to `jsKeywords`?

File: src/Language/PureScript/CodeGen/JS/Common.hs [hunk 2]
```diff
@@ -75,107 +81,160 @@ identCharToString c = '$' : show (ord c)
...
+  , "export"
+  , "extends"
+  , "finally"
+  , "for"
+  , "function"
+  , "if"
+  , "import"
+  , "in"
+  , "instanceof"
+  , "let"
```
Comment (michaelficarra) [inline src/Language/PureScript/CodeGen/JS/Common.hs]: Including `let` and `yield` in here is a bit inappropriate. I'd prefer to see `let`, `yield`, and `static` in their own list of `jsSometimesReserved`.

File: src/Language/PureScript/CodeGen/JS/Common.hs [hunk 3]
```diff
@@ -75,107 +81,160 @@ identCharToString c = '$' : show (ord c)
...
+  ]
+
+jsFutureReserved :: [String]
+jsFutureReserved =
+  [ "enum"
+  , "await"
+  ]
+
+jsFutureReservedStrict :: [String]
+jsFutureReservedStrict =
```
Comment (michaelficarra) [inline src/Language/PureScript/CodeGen/JS/Common.hs]: Can you sort this list and remove `static`?

File: src/Language/PureScript/CodeGen/JS/Common.hs [hunk 4]
```diff
@@ -75,107 +81,160 @@ identCharToString c = '$' : show (ord c)
...
+  , "package"
+  , "protected"
+  , "static"
+  , "interface"
+  , "private"
+  , "public"
+  ]
+
+jsOldReserved :: [String]
+jsOldReserved =
```
Comment (michaelficarra) [inline src/Language/PureScript/CodeGen/JS/Common.hs]: This list is missing `throws`.

File: src/Language/PureScript/CodeGen/JS/Common.hs (hunk 3 above)
Comment (garyb) [inline src/Language/PureScript/CodeGen/JS/Common.hs]: I found `static` [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Lexical_grammar#Future_reserved_keywords), is that incorrect then?

Comment (michaelficarra) [inline src/Language/PureScript/CodeGen/JS/Common.hs]: See the NOTE in [11.6.2.1](http://www.ecma-international.org/ecma-262/6.0/index.html#sec-keywords).

Comment (garyb) [inline src/Language/PureScript/CodeGen/JS/Common.hs]: Oh sorry, I see now that you wanted to move it to `jsSometimesReserved`, I thought you were suggesting removing it entirely.

**Comment by garyb** [comment]
@michaelficarra thanks for the feedback! I've updated the PR with your suggestions.

File: src/Language/PureScript/CodeGen/JS/Common.hs [hunk 5]
```diff
@@ -75,107 +81,166 @@ identCharToString c = '$' : show (ord c)
...
+    , "Symbol"
+    , "SyntaxError"
+    , "TypeError"
+    , "Uint16Array"
+    , "Uint32Array"
+    , "Uint8Array"
+    , "Uint8ClampedArray"
+    , "undefined"
+    , "unescape"
+    , "uneval"
```
Comment (michaelficarra) [inline src/Language/PureScript/CodeGen/JS/Common.hs]: Where is this one coming from?

File: src/Language/PureScript/CodeGen/JS/Common.hs [hunk 6]
```diff
@@ -75,107 +81,166 @@ identCharToString c = '$' : show (ord c)
...
+    , "parseFloat"
+    , "parseInt"
+    , "Promise"
+    , "Proxy"
+    , "RangeError"
+    , "ReferenceError"
+    , "Reflect"
+    , "RegExp"
+    , "Set"
+    , "SIMD"
```
Comment (michaelficarra) [inline src/Language/PureScript/CodeGen/JS/Common.hs]: Not official yet, but I'm okay with it.

File: src/Language/PureScript/CodeGen/JS/Common.hs [hunk 7]
```diff
@@ -75,107 +81,166 @@ identCharToString c = '$' : show (ord c)
...
+jsSometimesReserved :: [String]
+jsSometimesReserved =
+  [ "let"
+  , "static"
+  , "yield"
+  ]
+
+jsFutureReserved :: [String]
+jsFutureReserved =
+  [ "await"
```
Comment (michaelficarra) [inline src/Language/PureScript/CodeGen/JS/Common.hs]: Ugh, this is my fault for not noticing earlier. I would also move `await` to `jsSometimesReserved`, as it's only reserved in module code.

File: src/Language/PureScript/CodeGen/JS/Common.hs [hunk 8]
```diff
@@ -75,107 +81,166 @@ identCharToString c = '$' : show (ord c)
...
+    , "encodeURI"
+    , "encodeURIComponent"
+    , "Error"
+    , "escape"
+    , "eval"
+    , "EvalError"
+    , "Float32Array"
+    , "Float64Array"
+    , "Function"
+    , "Generator"
```
Comment (michaelficarra) [inline src/Language/PureScript/CodeGen/JS/Common.hs]: `Generator` and `GeneratorFunction` definitely shouldn't be in this list.

File: src/Language/PureScript/CodeGen/JS/Common.hs [hunk 9]
```diff
@@ -75,107 +81,166 @@ identCharToString c = '$' : show (ord c)
...
+    , "GeneratorFunction"
+    , "Infinity"
+    , "Int16Array"
+    , "Int32Array"
+    , "Int8Array"
+    , "InternalError"
```
Comment (michaelficarra) [inline src/Language/PureScript/CodeGen/JS/Common.hs]: This doesn't belong here either.

**Comment by michaelficarra** [comment]
LGTM otherwise.

File: src/Language/PureScript/CodeGen/JS/Common.hs (hunk 5 above)
Comment (garyb) [inline src/Language/PureScript/CodeGen/JS/Common.hs]: It's non-standard but always present in Firefox.

File: src/Language/PureScript/CodeGen/JS/Common.hs (hunk 9 above)
Comment (garyb) [inline src/Language/PureScript/CodeGen/JS/Common.hs]: Another FF global value...

File: src/Language/PureScript/CodeGen/JS/Common.hs (hunk 5 above)
Comment (michaelficarra) [inline src/Language/PureScript/CodeGen/JS/Common.hs]: Do we care about those? There's also plenty of non-standard globals in other browsers, such as `chrome` in Chrome and `opera` in Opera.

Comment (garyb) [inline src/Language/PureScript/CodeGen/JS/Common.hs]: Yeah, probably not; we'll never cover every possible non-standard value in every browser. I'll take them out.

**Comment by garyb** [comment]
Updated again. This should be good to merge now @paf31.

**Comment by michaelficarra** [comment]
:+1:

**Comment by paf31** [comment]
:+1: Sounds good.

**Pull request merged**